### DISCUSSION TITLE
blog/istio-sidecar-traffic-types/

### DISCUSSION THREAD

**Discussion by giscus** [rootsongjc/rootsongjc.github.io#287] [Announcements]
# blog/istio-sidecar-traffic-types/

本文将向你展示 Istio sidecar 中的六种流量类型及其 iptables 规则，并以示意图的形式带你一览其全貌。

https://jimmysong.io/blog/istio-sidecar-traffic-types/

**Comment by bigfish02** [comment]
请教一个问题，envoy 通过 127.0.0.6 访问业务容器，请求从 postrouting 出来后不应该是再次进入 prerouting 吗？为什么postrouting 出来后直接到业务容器了？

**Reply by rootsongjc** [reply]
如果再次进入 prerouting 的话就路由就循环了，请求从远程主机到达目标主机后只经过两次 iptables 就到达应用程序容器了，如下图所示：
![非同主机 Pod 间的服务访问路径（iptables 模式）](https://jimmysong.io/blog/beyond-istio-oss/iptables-process.svg)
详见：https://jimmysong.io/blog/beyond-istio-oss/#ebpf

**Comment by zhangmingcheng** [comment]
您好，我想请教下 iptables规则里面any包含lo网卡吗？

**Reply by zhangmingcheng** [reply]
网卡值为any包含lo网卡吗

**Reply by zhangmingcheng** [reply]
类型四：Local Pod -> Local Pod：Local Pod-> OUTPUT -> ISTIO_OUTPUT RULE 9   这里的RULE9总感觉不对

**Reply by rootsongjc** [reply]
any 应该包括 lo 网卡，我觉得，本机 pod 之间的互相访问，依然使用的是 IP 地址，而不是 localhost，因为是应用程序发起的访问所有进程的 owner 也不是 Envoy（即 UID/GID 1337），应该适用于 `ISTIO_RULE` 规则 9，你不知道你怎么看？

**Reply by zhangmingcheng** [reply]
您好，同一个Pod里面两个业务容器通过PodIp相互访问走的也是lo网卡吧

**Reply by Asongguo** [reply]
Pod里面两个业务容器互相访问，无论是用eth0网的ip还是127.0.0.1,都走lo网卡，可以抓包验证下。

**Comment by liyihuang** [comment]
Prometheus-> RREROUTING -> ISTIO_INBOUND（对目的地为 15002、15090 端口流量将转到 INPUT）-> INPUT -> Local Pod


这个地方应该是15020，15002 估计是笔误？

**Reply by rootsongjc** [reply]
确实是笔误，已修改。

**Comment by csgtree** [comment]
类型四Local Pod -> Local Pod的第3步，为什么不是rule 3呢？
`RETURN	any	lo	anywhere	anywhere !owner UID match 1337`
本地流量的out-interface应该是lo？

**Reply by rootsongjc** [reply]
你指的是这样的吗？
![图片](https://p.ipic.vip/3du2ay.png)

**Reply by csgtree** [reply]
是的。不知道有没有办法验证到底走了哪个flow

**Reply by liyihuang** [reply]
开个wireshark 再看着iptable的匹配就好了

**Reply by rootsongjc** [reply]
@csgtree 你的这个问题 @liyihuang 三个月前也提出过，我现在也不太能确定了。

**Reply by Asongguo** [reply]
测试了下，业务容器监听80端口，然后进入pod的网络空间，使用root用户，通过eth0网卡的ip发起请求：`curl 172.20.3.182:80`，在eth0网卡抓不到包，但是在lo网卡能抓到。同样`curl 127.0.0.1:80`也只能在lo网卡抓到包。

然后再通过iptables规则验证
root用户执行`curl 172.20.3.182:80`，通过`iptables -L -n -t nat -v`观察ISTIO_OUTPUT的RULE 3和RULE 9
请求前：
```
RULE3 27  1620 RETURN     all  --  *      lo      0.0.0.0/0            0.0.0.0/0            ! owner UID match 1337
RULE9 16   960 ISTIO_REDIRECT  all  --  *      *       0.0.0.0/0            0.0.0.0/0
```

请求后：
```
RULE3 28  1680 RETURN     all  --  *      lo      0.0.0.0/0            0.0.0.0/0            ! owner UID match 1337
RULE9 16   960 ISTIO_REDIRECT  all  --  *      *       0.0.0.0/0            0.0.0.0/0
```

RULE9处理的流量是960没变化，RULE3增加了60，说明流量确实走的RULE 3。

这里比较反直觉的是使用eth0网卡的ip请求本pod服务居然走lo网卡。

**Comment by Asongguo** [comment]
类型五：Envoy 内部的进程间 TCP 流量
Envoy 内部进程的 UID 和 GID 为 1337，它们之间的流量将使用 lo 网卡，使用 localhost 域名来通信。

这些流量通过的 iptables 规则如下。

Envoy 进程（Localhost） -> OUTPUT -> ISTIO_OUTPUT RULE 8 -> POSTROUTING -> Envoy 进程（Localhost）

RULE 4：
RETURN	any	any	anywhere	  anywhere  owner UID match 1337

RULE 8：
RETURN	any	any	anywhere	  localhost

这里为什么是 RULE 8而不是RULE 4呢？RULE 4也能匹配到这个流量吧？

**Reply by Asongguo** [reply]
去验证了下，在istio-proxy容器内，用id为1337的用户执行 `curl 127.0.0.1:15000`，然后执行`iptables -L -n -t nat -v`可以看到
`25  1500 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            owner UID match 1337`
变成
`26  1560 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            owner UID match 1337`

流量是到了RULE 4的

**Comment by qinlonglong123** [comment]
类型三：Prometheus -> Local Pod 
15020应该是在envoy进程监听的，4应该指向envoy进程吧
